File: src/fyi_system/process_events.py
"""Versioned, public-safe process-event export for fyi-archive consumers."""
# The exporter intentionally keeps validation and projection together so the
# public boundary is enforced at the serialization point.
# ruff: noqa: C901,EM101,EM102,PLC0415,PLW0108,TC003,TRY003,TRY004

from __future__ import annotations

import hashlib
import json
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

SCHEMA_VERSION = "1.0.0"
EXCLUDED_FIELDS = frozenset({"title", "request_title", "body", "detail", "excerpt", "requester"})


def _canonical(value: Any) -> bytes:
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode(
        "utf-8",
    )


def _sha256(value: Any) -> str:
    return hashlib.sha256(_canonical(value)).hexdigest()


def _utc(value: Any) -> tuple[str | None, str]:
    if not isinstance(value, str) or not value:
        return None, "missing"
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None, "invalid"
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=UTC)
    return parsed.astimezone(UTC).isoformat().replace("+00:00", "Z"), "valid"


def _request_id(row: dict[str, Any]) -> int:
    value = row.get("request_id", row.get("id"))
    if not isinstance(value, int) or value < 1:
        try:
            value = int(value)
        except (TypeError, ValueError) as exc:
            raise ValueError("request record must contain a positive integer id") from exc
    if value < 1:
        raise ValueError("request record must contain a positive integer id")
    return value


def _events(row: dict[str, Any]) -> list[dict[str, Any]]:
    """Use the source timeline when present; otherwise use the existing extractor."""
    direct = row.get("info_request_events")
    if isinstance(direct, list):
        return [item for item in direct if isinstance(item, dict)]
    from .fetch import extract_request_artifacts

    return extract_request_artifacts(row).get("events", [])


def _logical_request_id(request_id: int, instance_id: str) -> str:
    return f"urn:fyi:{instance_id}:request:{request_id}"


def _activity(event: dict[str, Any]) -> str:
    raw = (
        event.get("event_type")
        or event.get("type")
        or event.get("described_state")
        or event.get("state")
        or event.get("title")
        or "observed"
    )
    return str(raw).strip().lower().replace(" ", "_")


def _source_ref(event: dict[str, Any], index: int) -> str:
    return str(
        event.get("id") or event.get("event_id") or event.get("path") or f"source-index:{index}",
    )


def export_process_events(
    *,
    derived_dir: Path,
    output: Path,
    captured_at: str,
    checkpoint: Path | None = None,
    instance_id: str = "nz-fyi",
    source: str = "urn:fyi-cli:site:fyi.org.nz",
    base_url: str = "https://fyi.org.nz",
    attachments_output: Path | None = None,
) -> dict[str, Any]:
    """Emit deterministic NDJSON events and advance a resumable checkpoint."""
    if not derived_dir.is_dir():
        raise FileNotFoundError(f"derived request store does not exist: {derived_dir}")
    previous: dict[str, dict[str, Any]] = {}
    if checkpoint and checkpoint.exists():
        payload = json.loads(checkpoint.read_text(encoding="utf-8"))
        if not isinstance(payload, dict) or not isinstance(payload.get("event_digests", {}), dict):
            raise ValueError("checkpoint must contain an event_digests object")
        for key, value in payload["event_digests"].items():
            previous[str(key)] = (
                {"digest": str(value), "revision": 1} if isinstance(value, str) else value
            )

    rows = []
    for path in sorted(derived_dir.glob("*/*/request.json")):
        row = json.loads(path.read_text(encoding="utf-8"))
        if not isinstance(row, dict):
            raise ValueError(f"request record must be an object: {path}")
        rows.append(row)
    rows.sort(key=lambda row: _request_id(row))

    events: list[dict[str, Any]] = []
    attachments: list[dict[str, Any]] = []
    digests: dict[str, dict[str, Any]] = {}
    current_ids: set[str] = set()
    for row in rows:
        request_id = _request_id(row)
        logical_id = _logical_request_id(request_id, instance_id)
        source_events = _events(row)
        from .fetch import extract_request_artifacts

        source_attachments = row.get("attachments")
        if not isinstance(source_attachments, list):
            source_attachments = extract_request_artifacts(row).get("attachments", [])
        for index, raw in enumerate(source_events):
            source_ref = _source_ref(raw, index)
            event_id = f"{logical_id}:event:{_sha256((source_ref, index))}"
            timestamp, timestamp_status = _utc(
                raw.get("occurred_at") or raw.get("created_at") or raw.get("updated_at"),
            )
            item = {
                "schema_version": SCHEMA_VERSION,
                "event_id": event_id,
                "logical_request_id": logical_id,
                "activity": _activity(raw),
                "state": str(raw.get("state") or raw.get("described_state") or "") or None,
                "timestamp": timestamp,
                "timestamp_status": timestamp_status,
                "source_order": {
                    "source": source,
                    "request_sequence": request_id,
                    "event_sequence": index,
                },
                "provenance": {
                    "source_ref": source_ref,
                    "message_reference_id": str(
                        raw.get("message_id") or raw.get("message_reference_id") or "",
                    )
                    or None,
                    "capture_uri": f"{base_url.rstrip('/')}/request/{request_id}.json",
                    "captured_at": captured_at,
                },
            }
            digest = _sha256(item)
            old = previous.get(event_id)
            revision = int(old.get("revision", 0)) + 1 if old else 1
            item["revision"] = revision
            item["operation"] = "upsert"
            current_ids.add(event_id)
            digests[event_id] = {"digest": digest, "revision": revision}
            if old and not old.get("deleted") and old.get("digest") == digest:
                continue
            events.append(item)

        for attachment_index, attachment in enumerate(source_attachments):
            if not isinstance(attachment, dict):
                continue
            url = str(attachment.get("url") or "")
            attachment_id = f"{logical_id}:attachment:{_sha256((url, attachment_index))}"
            metadata = {
                "schema_version": SCHEMA_VERSION,
                "attachment_id": attachment_id,
                "logical_request_id": logical_id,
                "source_order": {
                    "source": source,
                    "request_sequence": request_id,
                    "attachment_sequence": attachment_index,
                },
                "content_type": str(attachment.get("content_type") or "") or None,
                "byte_size": attachment.get("size")
                if isinstance(attachment.get("size"), int)
                else None,
                "locator": {"uri": url} if url else None,
                "warc_record_id": row.get("warc_record_id") or row.get("warc_uri") or None,
                "provenance": {
                    "captured_at": captured_at,
                    "source_path": str(attachment.get("path") or ""),
                },
            }
            attachments.append(metadata)

    for event_id, old in sorted(previous.items()):
        if event_id not in current_ids and not old.get("deleted"):
            logical_id = event_id.split(":event:", 1)[0]
            events.append(
                {
                    "schema_version": SCHEMA_VERSION,
                    "event_id": event_id,
                    "logical_request_id": logical_id,
                    "activity": "removed",
                    "timestamp": captured_at,
                    "source_order": {"source": source, "event_sequence": len(events)},
                    "provenance": {"captured_at": captured_at},
                    "operation": "delete",
                    "revision": int(old.get("revision", 1)) + 1,
                },
            )
            digests[event_id] = {
                "digest": str(old.get("digest", "")),
                "revision": int(old.get("revision", 1)) + 1,
                "deleted": True,
            }

    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_bytes(b"".join(_canonical(item) + b"\n" for item in events))
    if attachments_output is not None:
        attachments_output.parent.mkdir(parents=True, exist_ok=True)
        attachments_output.write_bytes(b"".join(_canonical(item) + b"\n" for item in attachments))
    if checkpoint:
        checkpoint.parent.mkdir(parents=True, exist_ok=True)
        checkpoint.write_text(
            json.dumps(
                {
                    "schema_version": SCHEMA_VERSION,
                    "captured_at": captured_at,
                    "event_digests": digests,
                },
                sort_keys=True,
                indent=2,
            )
            + "\n",
            encoding="utf-8",
        )
    return {
        "schema_version": SCHEMA_VERSION,
        "event_count": len(events),
        "total_event_count": len(digests),
        "attachment_count": len(attachments),
        "output": str(output),
    }


def validate_public_event(event: dict[str, Any]) -> None:
    """Fail closed if an exported event contains a prohibited field."""

    def walk(value: Any) -> set[str]:
        if isinstance(value, dict):
            return set(value).intersection(EXCLUDED_FIELDS) | set().union(
                *(walk(item) for item in value.values()),
            )
        if isinstance(value, list):
            return set().union(*(walk(item) for item in value))
        return set()

    leaked = walk(event)
    if leaked:
        raise ValueError(f"public event contains excluded fields: {sorted(leaked)}")
    required = {
        "schema_version",
        "event_id",
        "logical_request_id",
        "activity",
        "timestamp",
        "source_order",
        "provenance",
    }
    missing = required.difference(event)
    if missing:
        raise ValueError(f"public event is missing fields: {sorted(missing)}")


def validate_process_event_file(path: Path) -> dict[str, int]:
    """Validate an exported NDJSON file and return its row count."""
    count = 0
    for line_number, line in enumerate(path.read_text(encoding="utf-8").splitlines(), start=1):
        if not line.strip():
            continue
        try:
            value = json.loads(line)
        except json.JSONDecodeError as exc:
            raise ValueError(f"invalid JSON at line {line_number}") from exc
        if not isinstance(value, dict):
            raise TypeError(f"row at line {line_number} must be an object")
        if value.get("schema_version") != SCHEMA_VERSION:
            raise ValueError(f"unsupported schema version at line {line_number}")
        validate_public_event(value)
        count += 1
    return {"row_count": count}

File: crates/fyi-core/src/process_events.rs
//! Public-safe, deterministic process-event export for archive consumers.

use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};
use thiserror::Error;

pub const SCHEMA_VERSION: &str = "1.0.0";

#[derive(Debug, Error)]
pub enum ProcessEventError {
    #[error("derived request store does not exist: {0}")]
    MissingStore(PathBuf),
    #[error("invalid JSON in {path}: {source}")]
    Json {
        path: PathBuf,
        source: serde_json::Error,
    },
    #[error("request record must be an object: {0}")]
    InvalidRecord(PathBuf),
    #[error("checkpoint must contain an event_digests object")]
    InvalidCheckpoint,
    #[error("request record must contain a positive integer id")]
    InvalidRequestId,
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),
    #[error("serialization error: {0}")]
    Serialization(#[from] serde_json::Error),
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ExportSummary {
    pub schema_version: String,
    pub event_count: usize,
    pub total_event_count: usize,
    pub attachment_count: usize,
}

fn digest(value: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(value.as_bytes());
    format!("{:x}", hasher.finalize())
}

fn request_id(row: &Map<String, Value>) -> Result<i64, ProcessEventError> {
    row.get("request_id")
        .or_else(|| row.get("id"))
        .and_then(Value::as_i64)
        .filter(|value| *value > 0)
        .ok_or(ProcessEventError::InvalidRequestId)
}

fn collect_request_files(root: &Path, output: &mut Vec<PathBuf>) -> Result<(), ProcessEventError> {
    if !root.is_dir() {
        return Ok(());
    }
    for entry in fs::read_dir(root)? {
        let path = entry?.path();
        if path.is_dir() {
            collect_request_files(&path, output)?;
        } else if path.file_name().and_then(|name| name.to_str()) == Some("request.json") {
            output.push(path);
        }
    }
    Ok(())
}

fn event_rows(row: &Map<String, Value>) -> Vec<&Map<String, Value>> {
    if let Some(Value::Array(events)) = row.get("info_request_events") {
        return events.iter().filter_map(Value::as_object).collect();
    }
    let mut found = Vec::new();
    fn walk<'a>(value: &'a Value, path: &str, found: &mut Vec<&'a Map<String, Value>>) {
        match value {
            Value::Object(map) => {
                let eventish = path.contains("event")
                    || path.contains("history")
                    || path.contains("message")
                    || map.contains_key("event_type");
                if eventish && (map.contains_key("created_at") || map.contains_key("event_type")) {
                    found.push(map);
                }
                for (key, child) in map {
                    walk(child, &format!("{path}.{key}"), found);
                }
            }
            Value::Array(items) => {
                for (index, child) in items.iter().enumerate() {
                    walk(child, &format!("{path}[{index}]"), found);
                }
            }
            _ => {}
        }
    }
    for (key, value) in row {
        walk(value, key, &mut found);
    }
    found
}

fn activity(event: &Map<String, Value>) -> String {
    event
        .get("event_type")
        .or_else(|| event.get("type"))
        .or_else(|| event.get("described_state"))
        .or_else(|| event.get("state"))
        .and_then(Value::as_str)
        .unwrap_or("observed")
        .trim()
        .to_lowercase()
        .replace(' ', "_")
}

fn source_ref(event: &Map<String, Value>, index: usize) -> String {
    event
        .get("id")
        .or_else(|| event.get("event_id"))
        .and_then(Value::as_str)
        .map(String::from)
        .unwrap_or_else(|| format!("source-index:{index}"))
}

fn load_checkpoint(path: Option<&Path>) -> Result<BTreeMap<String, Value>, ProcessEventError> {
    let Some(path) = path else {
        return Ok(BTreeMap::new());
    };
    if !path.exists() {
        return Ok(BTreeMap::new());
    }
    let value: Value = serde_json::from_str(&fs::read_to_string(path)?)?;
    let Some(events) = value.get("event_digests").and_then(Value::as_object) else {
        return Err(ProcessEventError::InvalidCheckpoint);
    };
    Ok(events
        .iter()
        .map(|(key, value)| (key.clone(), value.clone()))
        .collect())
}

pub fn export_process_events(
    derived_dir: &Path,
    output: &Path,
    captured_at: &str,
    checkpoint_path: Option<&Path>,
    attachments_output: Option<&Path>,
) -> Result<ExportSummary, ProcessEventError> {
    if !derived_dir.is_dir() {
        return Err(ProcessEventError::MissingStore(derived_dir.to_path_buf()));
    }
    let previous = load_checkpoint(checkpoint_path)?;
    let mut paths = Vec::new();
    collect_request_files(derived_dir, &mut paths)?;
    paths.sort();
    let mut rows = Vec::new();
    for path in paths {
        let value: Value = serde_json::from_str(&fs::read_to_string(&path)?)?;
        let Some(row) = value.as_object() else {
            return Err(ProcessEventError::InvalidRecord(path));
        };
        rows.push((request_id(row)?, row.clone()));
    }
    rows.sort_by_key(|(id, _)| *id);

    let mut events = Vec::new();
    let mut attachments = Vec::new();
    let mut current = BTreeSet::new();
    let mut digests = BTreeMap::new();
    for (request_id, row) in rows {
        let logical_id = format!("urn:fyi:nz-fyi:request:{request_id}");
        for (index, raw) in event_rows(&row).into_iter().enumerate() {
            let reference = source_ref(raw, index);
            let event_id = format!(
                "{logical_id}:event:{}",
                digest(&format!("{reference}:{index}"))
            );
            current.insert(event_id.clone());
            let mut event = json!({
                "schema_version": SCHEMA_VERSION,
                "event_id": event_id.clone(),
                "logical_request_id": logical_id.clone(),
                "activity": activity(raw),
                "timestamp": raw.get("occurred_at")
                    .or_else(|| raw.get("created_at"))
                    .or_else(|| raw.get("updated_at"))
                    .and_then(Value::as_str)
                    .unwrap_or(captured_at),
                "source_order": {
                    "source": "urn:fyi-cli:site:fyi.org.nz",
                    "request_sequence": request_id,
                    "event_sequence": index
                },
                "provenance": {"source_ref": reference, "captured_at": captured_at},
                "operation": "upsert"
            });
            let event_digest = digest(&serde_json::to_string(&event)?);
            let old = previous.get(event_id.as_str());
            let revision = old
                .and_then(|value| value.get("revision"))
                .and_then(Value::as_u64)
                .unwrap_or(0)
                + 1;
            event["revision"] = json!(revision);
            let unchanged = old
                .filter(|value| value.get("deleted").and_then(Value::as_bool) != Some(true))
                .and_then(|value| value.get("digest"))
                .and_then(Value::as_str)
                == Some(event_digest.as_str());
            digests.insert(
                event_id,
                json!({"digest": event_digest, "revision": revision}),
            );
            if !unchanged {
                events.push(event);
            }
        }
        if let Some(Value::Array(items)) = row.get("attachments").or_else(|| row.get("files")) {
            for (index, item) in items.iter().filter_map(Value::as_object).enumerate() {
                attachments.push(json!({
                    "schema_version": SCHEMA_VERSION,
                    "attachment_id": format!(
                        "{logical_id}:attachment:{}",
                        digest(&format!("{}:{index}", item.get("url").and_then(Value::as_str).unwrap_or("")))
                    ),
                    "logical_request_id": logical_id.clone(),
                    "source_order": {
                        "source": "urn:fyi-cli:site:fyi.org.nz",
                        "request_sequence": request_id,
                        "attachment_sequence": index
                    },
                    "content_type": item.get("content_type").cloned().unwrap_or(Value::Null),
                    "byte_size": item.get("size").cloned().unwrap_or(Value::Null),
                    "locator": item.get("url").map(|url| json!({"uri": url})).unwrap_or(Value::Null),
                    "warc_record_id": row.get("warc_record_id").cloned().unwrap_or(Value::Null),
                    "provenance": {"captured_at": captured_at, "source_path": "request.json"}
                }));
            }
        }
    }
    for (event_id, old) in &previous {
        if !current.contains(event_id) && old.get("deleted").and_then(Value::as_bool) != Some(true)
        {
            let logical_id = event_id.split(":event:").next().unwrap_or(event_id);
            let revision = old.get("revision").and_then(Value::as_u64).unwrap_or(1) + 1;
            events.push(json!({
                "schema_version": SCHEMA_VERSION,
                "event_id": event_id,
                "logical_request_id": logical_id,
                "activity": "removed",
                "timestamp": captured_at,
                "source_order": {
                    "source": "urn:fyi-cli:site:fyi.org.nz",
                    "request_sequence": 0,
                    "event_sequence": events.len()
                },
                "provenance": {"captured_at": captured_at},
                "operation": "delete",
                "revision": revision
            }));
            digests.insert(
                event_id.clone(),
                json!({"digest": old.get("digest").cloned().unwrap_or(Value::Null), "revision": revision, "deleted": true}),
            );
        }
    }
    write_ndjson(output, &events)?;
    if let Some(path) = attachments_output {
        write_ndjson(path, &attachments)?;
    }
    if let Some(path) = checkpoint_path {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::write(
            path,
            serde_json::to_string_pretty(&json!({
                "schema_version": SCHEMA_VERSION,
                "captured_at": captured_at,
                "event_digests": digests
            }))? + "\n",
        )?;
    }
    Ok(ExportSummary {
        schema_version: SCHEMA_VERSION.to_string(),
        event_count: events.len(),
        total_event_count: digests.len(),
        attachment_count: attachments.len(),
    })
}

fn write_ndjson(path: &Path, rows: &[Value]) -> Result<(), ProcessEventError> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let mut output = String::new();
    for row in rows {
        output.push_str(&serde_json::to_string(row)?);
        output.push('\n');
    }
    fs::write(path, output)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::{SystemTime, UNIX_EPOCH};

    #[test]
    fn exports_source_order_and_redacts_attachment_names() {
        let root = std::env::temp_dir().join(format!(
            "fyi-process-events-{}",
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .expect("clock")
                .as_nanos()
        ));
        let store = root.join("Agency").join("42");
        fs::create_dir_all(&store).expect("store");
        fs::write(
            store.join("request.json"),
            r#"{"id":42,"title":"private","info_request_events":[{"id":"a","event_type":"closed"},{"id":"b","event_type":"opened"}],"files":[{"name":"private.pdf","url":"https://fyi.example/attach/1","size":12}]}"#,
        )
        .expect("request");
        let events = root.join("events.ndjson");
        let attachments = root.join("attachments.ndjson");
        let summary = export_process_events(
            &root.join("Agency"),
            &events,
            "2026-01-01T00:00:00Z",
            None,
            Some(&attachments),
        )
        .expect("export");
        assert_eq!(summary.event_count, 2);
        let rows: Vec<Value> = fs::read_to_string(&events)
            .expect("events")
            .lines()
            .map(|line| serde_json::from_str(line).expect("json"))
            .collect();
        assert_eq!(rows[0]["source_order"]["event_sequence"], 0);
        assert_eq!(rows[1]["source_order"]["event_sequence"], 1);
        assert!(!fs::read_to_string(&attachments)
            .expect("attachments")
            .contains("private.pdf"));
        let _ = fs::remove_dir_all(root);
    }
}
